lib/selection.js

```diff
@@ -0,0 +1,202 @@
+      const existing = this._items[i]
+
+      if (existing.isStreamSelection) {
+        if (isStreamSelection) {
+          // If both are stream selections and they match, then we remove the first matching item, then we break the loop
+          if (existing.from === item.from && existing.to === item.to) {
+            this._items.splice(i, 1)
+            // for stream selections, we only remove one item at a time
+            // ergo we break the loop after removing the first matching item
+            break
+          }
+        } else {
+          // we only remove stream selections when the `isStreamSelection` flag is true and they match
+          continue
+        }
+      }
+
+      if (isLowerIntersecting(item, existing)) {
+        existing.to = item.from - 1
+        continue
+      }
+      if (isUpperIntersecting(item, existing)) {
+        existing.from = item.to + 1
+        continue
+      }
+      if (isInsideExisting(item, existing)) {
+        const existingStart = { ...existing, to: item.from - 1 }
+        const existingEnd = { ...existing, from: item.to + 1 }
+        this._items.splice(i, 1, existingStart, existingEnd)
+        continue
+      }
+      if (isCoveringExisting(item, existing)) {
+        this._items.splice(i, 1)
+        i -= 1
+        continue
+      }
+    }
+  }
+
+  /**
+   * @public
+   * @param {SelectionItem} newItem
+   * @param {boolean} [isStreamSelection]
+   */
+  insert (newItem, isStreamSelection = false) {
+    if (newItem.from > newItem.to) {
+      throw new Error('Invalid interval')
+    }
+    if (!isStreamSelection) {
+      this.remove(newItem)
+    }
+    this._items.push({ ...newItem, isStreamSelection })
+  }
+
+  /**
+   * @public
+   * @param {(a: SelectionItem, b: SelectionItem) => number} sortFn
+   */
+  sort (sortFn = (a, b) => a.from - b.from) {
+    this._items.sort(sortFn)
+  }
+
+  /**
+   * @public
+   */
+  isEmpty () {
+    return this._items.length === 0
+  }
+
+  /**
+   * @public
+   */
+  get length () {
+    return this._items.length
+  }
+
+  /**
+   * Takes an index and returns the item at that index
+   * @public
+   * @param {number} index
+   */
+  get (index) {
+    return this._items[index]
+  }
+
+  /**
+   * Takes an index and an item and replaces the item at the index with the new item
+   * @public
+   * @param {number} index
+   * @param {SelectionItem} item
+   */
+  set (index, item) {
+    this._items[index] = item
+  }
+
+  /**
+   * Takes two indexes and swaps the items at those indexes
+   * @public
+   * @param {number} i
+   * @param {number} j
+   */
+  swap (i, j) {
+    const temp = this._items[i]
+    this._items[i] = this._items[j]
+    this._items[j] = temp
+  }
+
+  /**
+   * @public
+   */
+  clear () {
+    this._items.length = 0
+  }
+
+  * [Symbol.iterator] () {
+    const self = this
+    for (let i = 0; i < this._items.length; i++) {
+      const item = this._items[i]
+      yield {
```

Comment:
this.... works? but feels a bit too hacky? if u figure out a cleaner way to do this then great, if not oh well it's fine

Reply:
I don't consider this to be too hacky to be honest. The other solution would be to call `.splice` on `_selections._items` from outside of this file, which does feel hacky tbh.
Plus this allows us to use `for (.. of ..)` loops and still remove items without having to deal with the indexes. If you think of a better solution, I'd be happy to replace the current one, but I also don't really see this as an issue either, since it's easily understandable and simple.